### PR TITLE
Added macarthur_core/lovd module to project.

### PR DESCRIPTION
**Pull request by andrewhill157**
This additional code within leiden_sc/ contains:

1) A set of classes in macarthur_core/lovd/leiden_database.py:

These classes allow a user to extract tables of data (mutations listed for a specific gene in the database) and other useful information from any Leiden Open Variation Database installation, such as http://www.dmd.nl/nmdb2/. Unfortunately, it has been necessary to do this by downloading the HTML for relevant pages on the database and parsing out the necessary data, as they do not have an easy way to access the data otherwise. Therefore, I have added an external dependency - beautifulsoup4 - for HTML parsing. 

Basically, the usage for these classes goes like this:

```
leiden_url = 'http://www.dmd.nl/nmdb2/'
gene_id = 'ACTA1'

database = make_leiden_database(leiden_url)
database.set_gene_id(gene_id)
column_labels = leiden_database.get_table_headers()
table_entries = leiden_database.get_table_data()
...
```

Note that make_leiden_database acts as a factory method to generate the right subclass of LeidenDatabase for the detected version number. 

2) A set of functions in macarthur_core/lovd/utilities.py:

These are general utility functions, some of which are used in leiden_database.py.

### CODE REVIEW

File: leiden_sc/macarthur_core/lovd/leiden_database.py [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+        """
+        Must call set_gene_id prior to use.
+        Constructs URL linking to the table of variant entries for the specified gene_id on the Leiden Database site.
+
+        @return: URL linking to the table of variant entries for the specified gene_id on the
+        Leiden Database site.
+        @rtype: string
+        """
+
+        raise Exception("ABSTRACT METHOD NOT IMPLEMENTED")
```
Comment (zonca) [inline leiden_sc/macarthur_core/lovd/leiden_database.py:144]: you could raise `NonImplementedError` instead of `Exception`

File: leiden_sc/macarthur_core/lovd/leiden_database.py [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+    def __init__(self, leiden_url):
+
+        # Call to the super class constructor
+        LeidenDatabase.__init__(self, leiden_url)
+        self.version_number = 3
+
+        if not leiden_url.lower().endswith('/'):
+            leiden_url += '/'
+
+        if leiden_url.lower().endswith('genes/'):
```
Comment (zonca) [inline leiden_sc/macarthur_core/lovd/leiden_database.py:405]: I wonder whether some of this URL operations might be easier with `urlparse` http://docs.python.org/2/library/urlparse.html

File: leiden_sc/macarthur_core/lovd/utilities.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    @rtype: string
+    @raise: ValueError if there is no 4+ digit number in link_url
+    """
+
+    # Search for sequences of digits that are four digits or longer in length.
+    m = re.compile('\d{4,}')
+    results = m.search(link_url)
+
+    # Return entire matched sequence (PMID)
+    if results is not None:
```
Comment (zonca) [inline leiden_sc/macarthur_core/lovd/utilities.py:30]: probably `if results:` is enough

**Comment by zonca** [comment]
Thanks @andrewhill157 
first of all, this is very good code, it is clear, pep8, docstrings, good method and variable names.
I like the usage of a base class and a factory method to generate the correct derived class.

The only thing I don't like much is `set_gene_id` and the fact that all methods require `set_gene_id` to be called before.
At this point wouldn't it be simpler to just use the `gene_id` as input to the factory?
it looks to me that the process of building the class is cheap, so I would rather create a brand new instance for each gene, but maybe I'm missing something...

Do you also have some unit tests for this module?

In my understanding this looks like a pretty isolated module that could be useful by itself to scientists, I think it would be suitable to be published as a single package on `pip` instead of being a part of this project. What do you think?

**Comment by andrewhill157** [comment]
Thanks very much for taking the time to look at this @zonca ! I agree with all of your code comments and will look to incorporate them moving forward. 

My reasoning for using set_gene the way I did was that one or two of the methods are are not specific to a gene, but the database as a whole. For example, get_available_genes is really an operation on the database, not a specific gene. Users would potentially need this to find out what genes they can set before using set_gene_id (or constructing a LeidenDatabase object with a specific gene ID as you suggested). set_gene_id is currently a little slow because it has to download HTML from a few different pages in series. One option would be to separate functions that do not involve operations on gene pages (like get_available_genes) from the class and provide them as functions like make_leiden_database. I could then make the constructor include the gene ID as you describe. I agree, I am not particularly happy with how this currently works, and any input is very much appreciated!

I do have unit tests for this module; I will include them in the next PR along with one other (very small) module that will be needed. There are some functions that do not have unit tests. I am not particularly experienced in developing unit tests, so I am not sure how to develop good tests for some functions in this module. I would love to discuss unit testing strategies once I submit the PR.

This module is well isolated and would probably be a good candidate for a pip package. I would certainly be open to doing this for any parts of this project that might fit into the category.

**Comment by andrewhill157** [comment]
@zonca I am not sure how, but somehow my new commit got merged into this PR before I closed and merged this one. I am not sure how to cleanly reverse this process. 

Maybe the best thing is to comment on the additional commit in this pull request? I have added the unit tests for the relevant modules in the folder macarthur_core/test/. The test modules mirror the structure of the classes they test. I have also added one additional module macarthur_core/io/web_io.py, which is required to run the leiden_database.py code.

Does this sound like a viable plan? If you have a better strategy, please be sure to let me know.

**Comment by zonca** [comment]
Those commits look related, so let's continue on this PR.

**Comment by zonca** [comment]
About set_gene, what about extracting those 2 methods as separate functions? Possibly you can factor out some of the code needed in independent functions.
I think this would be easier to use, as most methods in the class are gene specific, it makes sense to have the object as a whole to be gene specific, even at the cost of duplicating some small operations in the class constructor and in this extracted functions.

**Comment by zonca** [comment]
Other option would be:
- create another level of classes, something like `GeneData`
- move all gene specific methods there
- the database objects create and return those objects with a method `get_gene_data(gene_id)`
- the `GeneData` object has a reference to the database object for calling its methods

anyway, you probably have large amount of code that depends on how this module is implemented now, so let's just go ahead and open a brand new PR for another package and work there.
